crates/adapters/src/controller/validate.rs

```diff
@@ -0,0 +1,162 @@
+use std::collections::{BTreeMap, HashSet};
+
+use feldera_adapterlib::errors::controller::ConfigError;
+use feldera_types::config::PipelineConfig;
+
+pub fn validate_config(config: &PipelineConfig) -> Result<(), ConfigError> {
+    let mut dependencies = Vec::new();
+
+    for (endpoint_name, input) in config.inputs.iter() {
+        if let Some(start_after) = input.connector_config.start_after.as_ref() {
+            if start_after.is_empty() {
+                return Err(ConfigError::empty_start_after(endpoint_name));
+            }
+            for start_after in start_after.iter() {
+                for label in input.connector_config.labels.iter() {
+                    dependencies.push((
+                        endpoint_name.to_string(),
+                        label.clone(),
+                        start_after.clone(),
+                    ));
+                }
+            }
+        }
+    }
+
+    // check for cycles
+    if let Some(cycle) = find_cycle(&dependencies) {
+        return Err(ConfigError::cyclic_dependency(cycle));
+    }
+
+    Ok(())
+}
+
+/// Detect the first cycle in a directed graph
+fn find_cycle(edges: &[(String, String, String)]) -> Option<Vec<(String, String)>> {
+    let mut graph: BTreeMap<String, Vec<(String, String)>> = BTreeMap::new();
+
+    // Build adjacency list from edge list
+    for (endpoint, u, v) in edges {
+        graph
+            .entry(u.clone())
+            .or_default()
+            .push((endpoint.clone(), v.clone()));
+    }
+
+    let mut visited = HashSet::new();
+    let mut recursion_stack = Vec::new();
+    let mut stack_set = HashSet::new(); // For quick lookup in the stack
+
+    fn dfs(
+        node: &str,
+        graph: &BTreeMap<String, Vec<(String, String)>>,
+        visited: &mut HashSet<String>,
+        recursion_stack: &mut Vec<(String, String)>,
+        stack_set: &mut HashSet<String>,
+    ) -> Option<Vec<(String, String)>> {
```

Comment:
don't we already have code for doing this somewhere?

Reply:
Not that I know of, and I didn't want to add another external dependency just for this.